File: programmers/모두 0으로 만들기.java
// long으로 반환해야 한다면 중간에 값이 long 범위인데 int로 할당하지 않았는지 잘 확인해야 한다.

import java.util.ArrayList;
import java.util.LinkedList;

class Solution {
    static ArrayList<Integer>[] list;
    static boolean[] check;
    static long answer = 0;
    static long[] aa;

    public long solution(int[] a, int[][] edges) {
        int[] temp = checkSum(a);
        if (temp[1] != 0) {
            return 0;
        } else if (temp[0] != 0) {
            return -1;
        }

        check = new boolean[a.length];
        aa = new long[a.length];
        list = new ArrayList[a.length];
        for (int i = 0; i < list.length; i++) {
            list[i] = new ArrayList<>();
            aa[i] = a[i];
        }
        for (int[] edge : edges) {
            int from = edge[0];
            int to = edge[1];
            list[from].add(to);
            list[to].add(from);
        }

        dfs(0, -1);

        return answer;
    }

    public void dfs (int current, int parent) {
        check[current] = true;

        for (int i = 0; i < list[current].size(); i++) {
            if (!check[list[current].get(i)]) {
                dfs(list[current].get(i), current);
            }
        }
        answer += Math.abs(aa[current]);
        if (parent != -1) {
            aa[parent] += aa[current];
        }
    }

    public int[] checkSum (int[] a) {
        int sum = 0;
        int check = 0;
        boolean allZero = true;

        for (int i = 0; i < a.length; i++) {
            if (a[i] != 0) {
                sum += a[i];
                allZero = false;
            }
        }

        if (allZero) {
            check = 1;
        }
        return new int[] {sum, check};
    }

//    public static void main (String[] arg) {
//        int[][] edges = {{0,1},{3,4},{2,3},{0,3}};
//        int[] a = {-5,0,2,1,2};
//        Solution s = new Solution();
//        System.out.println(s.solution(a,edges));
//    }
}

File: programmers/모두 0으로 만들기.swift
// dfs시 검사를 dfs 함수에 넘겨준 다음 실행하면 시간이 너무 오래걸린다.
// check 배열을 사용할 경우 매 dfs마다 새로 var check = check로 할당하는 경우도
// 시간을 오래 잡아먹었다. inout을 통해 reference로 접근하자.

import Foundation

func solution(_ a:[Int], _ edges:[[Int]]) -> Int64 {
    var tree: [[Int]] = [[Int]] (repeating: [], count: a.count)
    var answer: Int64 = 0
    var a = a
    var check = [Bool] (repeating: false, count: a.count)
    
    for edge in edges {
        tree[edge[0]].append(edge[1])
        tree[edge[1]].append(edge[0])
    }
    
    dfs(tree: tree, a: &a, current: 0, parent: -1, check: &check, answer: &answer)
    
    for i in a {
        if i != 0 {
            answer = -1
            break
        }
    }
    
    return answer
}

func dfs (tree: [[Int]], a: inout [Int], current: Int, parent: Int, check: inout [Bool], answer: inout Int64) {
    if check[current] { return }
    check[current] = true
    
    for next in tree[current] {
        if !check[next] {
            dfs(tree: tree, a: &a, current: next, parent: current, check: &check, answer: &answer)
        }
    }
    
    answer += Int64(abs(a[current]))
    if parent != -1 {
        a[parent] += a[current]
        a[current] = 0
    }
}
